mountinfo/mounted_linux_test.go

```diff
@@ -314,6 +322,38 @@ func TestMountedBy(t *testing.T) {
 				}
 			}
 
+			// Check the public MountedFast() function as a whole.
+			mounted, sure, err := MountedFast(m)
+			if err == nil {
+				if !openat2Supported {
+					// stat does not detect bind-mounts correctly.
+					if tc.isBind && mounted == true {
+						t.Errorf("MountFast: expected mounted as false for bind-mounts, got %v", mounted)
+					}
+					if !tc.isBind && mounted != exp {
+						t.Errorf("MountFast: expected mounted as %v, got %v", exp, mounted)
+					}
+					if sure != tc.isSure {
```

Comment:
I understand that this code is tests for `isSure == false` case in openat2 unsupported environment.
According to comment in `mountedFast`, `isSure == false` can be returned for `isMount == false` case or `isBind == true` case.
Therefore, expected `isSure` will be like below:

| openat2Supported  | isMount |  isBind |  expected isSure |    Note                          |
|------------------ |---------| --------| -----------------| ---------------------------------|
|  true             |  true   |   true  |    true          |                                  |
|  true             |  true   |   false |    true          |                                  |
|  true             |  false  |   true  |    true          |  This won't actually happen.     |
|  true             |  false  |   false |    true          |                                  |
|  false            |  true   |   true  |    false         |                                  |
|  false            |  true   |   false |    true          |                                  |
|  false            |  false  |   true  |    false         |  This won't actually happen.     |
|  false            |  false  |  false  |   false          |                                  |

Then, isn't it less confusing not defining `isSure` in each test case, instead making expected `isSure` by using `isMount` and `isBind`? I guess that it can be `isMount && !isBind` when `openat2Supported == false`. Manually defining `isSure` would make it difficult for other people to add additional tests.

Reply:
Hmm! I think adding `isSure` helps a bit and the test-writer should be able to know if the value they are adding is guaranteed to be a mount-point or not.

There are a couple of cases where `sure` is `false` for `openat2Supported` as well.
For eg: `non-existent path` and `broken_symlink`. Both of them lie in the path of `normalizePath` where is sure is returned as false which is the 3rd row.

I prefer keeping it as it is rather than increasing the branching as this test is really quite complicated.

Reply:
Oh, I missed to add the error return cases in the table. Thank you for pointing it out.
However, aren't they handled in else clause for `if err == nil`?

Actually, `sure` seems directly compared with `true` and `false` in else clauses just below, it would be confusing to keep `isSure` in the current way. So, if we decide to keep `isSure`, I prefer separately defining something like `isSureNoOpenat2` and `expectMountNoOpenat2` and comparing them with the actual returned values for `openat2Supported == false` case (We may also need to rename `isSure` to `isSureOpenat2` to avoid confusion and compare it for `openat2Supported == true` case).

BTW, `isSure: true` for "non-existent path" case seems incorrect, then.

Reply:
Yes, it makes sense to separate out `sure tests` for better readability. Yes, they are added there and that's how I realized these cases. I will try to do that.

For the case of `non-existent path`; The stat call leads to the failure: https://github.com/moby/sys/blob/26b84be33cb40ce258b8c616d0a0aa8c250ae356/mountinfo/mounted_unix.go#L41-L44 in normalizePath and hence we are not sure. Its the same as what k8s does today: Ref: https://github.com/kubernetes/kubernetes/blob/34d1d2a4b98e496145d49aee322a94d8387bfdf7/staging/src/k8s.io/mount-utils/mount_linux.go#L337-L340

FWIW, I think the user should not use the `MountedFast` if they want to be absolutely be sure.  They should choose `Mounted`.

Reply:
Thank you for your fix. It improved a lot.

I think that we can improve a bit by moving `openat2Supported` check to the top level like below (I'm not sure that it's correct, so please double-check):

```golang
// for any error, regardless of kernel version, sure is false and mounted is false.
if err != nil {
	if sure == true {}
	if mounted != false {}
}

if openat2Supported {
	if err == nil {
		if sure != true {}
		if mounted != exp {}
	}
	
	if tc.isMount && sure == false {} 
} else {
	if err == nil {
		// stat does not detect bind-mounts correctly.
		if tc.isBind && mounted == true {}
		if !tc.isBind && mounted != exp {}
	} 

	if tc.isMount {
		if tc.isBind && sure {} 
		if !tc.isBind && !sure {}
	} else {
		// if it is not a mount-point, we are never sure on older kernels.
		if sure {}
	}
}
```

By above refactoring, we can separately focus on the behaviors of openat2Supported case and non-openat2Supported case (Actually, I didn't noticed that we can simplify like it, until I see your refactored code. Very nice work).

Also, `tc.isSure` seems no longer referenced, can we remove `isSure` entirely?

Reply:
Thank you for handling this.

Should we also make error if `tc.isMount && !sure` and `!tc.isMount && sure` for `err == nil && !openat2Supported && !tc.isBind` case, or line 345?
(Actually, equivalent codes existed in the previous code).

These conditions are exactly what consumers like k8s will rely on whether to use fall back logic, it would be worth checking here. Other than this, it looks good to me.

Reply:
Sorry! I missed it; my bad; thank you for reviewing so patiently. I need to get better at keeping all the branchings; if you have suggestions I can take it as well. 

I will update it.

Reply:
Thank you. It looks good to me now.